### DISCUSSION TITLE
Is there a way to determine whether a program/kernel is running in dosemu -dumb as opposed to -t, SDL, or X?

### DISCUSSION THREAD

**Discussion by ecm-pushbx** [dosemu2/dosemu2#1411] [Q&A]
I'm considering adding colour/highlighting to my debugger's output. If the display is written to a serial port or using dosemu -dumb `int 10h` then I want to use ANSI escape sequences which will be handled by the host terminal. If any other dosemu mode (including -t) is in use then `int 10h` functions that write with colour attributes should be used. (If the debugger is used as a DOS application then I'll probably also use escape sequences and require the user to have installed an ANSI parser either in DOS or on the host terminal.)

Is there a way for a program (possibly bootloaded) to determine which default to use, that is determine whether -dumb is in effect?

**Comment by stsp** [comment]
Not that I am aware of.
But running -t over serial port should
work too, so I am not sure there is a
problem to begin with.

**Reply by ecm-pushbx** [reply]
Yes, I am not concerned about the serial I/O mode. Would just be neat for the non-serial case.

**Reply by stsp** [reply]
What I mean is that you should always
use int10. If there is some terminal over
the serial port, then -t mode can use it,
so why would you code that into debugger?

**Reply by ecm-pushbx** [reply]
By serial port I mean that the debugger application connects to the emulated serial port of the dosemu, generally connecting to a different terminal on the same host machine. I don't mean any hardware serial ports across which to use Linux. I'm not sure what you mean though.

**Reply by stsp** [reply]
But emulated serial port is either
connected to a real serial port or
to a pty. And in both cases you can
have some terminal renderer on the
other side (like xterm).
Which is I think what you were saying.
And dosemu -t can have such serial
port or a pty as a master terminal
rather than just an emulated serial
port.

**Reply by andrewbird** [reply]
I think @ecm-pushbx wants to write different output to screen: ANSI escapes if dumb,  else dos colour attributes if rendered by dosemu.

**Reply by stsp** [reply]
He is saying that:
```
If the display is written to a serial port or using dosemu -dumb then I want to
use ANSI escape sequences which will be handled by the host terminal.
```

And I am saying that there is no reason
to use dumb, even if you work with terminal
over a serial port.

**Reply by stsp** [reply]
s-lang supports variety of terminals,
including ANSI. So why to use -dumb
and hard-coding some terminal protocol
into the debuger? What's the use?

**Reply by andrewbird** [reply]
okay, I thought you'd missed the point about changing the output.

**Reply by ecm-pushbx** [reply]
> He is saying that:

*She

> But emulated serial port is either
> connected to a real serial port or
> to a pty.

It's connected to a pty created by socat and the other end is opened by picocom.

> And in both cases you can
> have some terminal renderer on the
> other side (like xterm).
> Which is I think what you were saying.
> And dosemu -t can have such serial
> port or a pty as a master terminal
> rather than just an emulated serial
> port.

Except for small tests where the set up is too much effort I generally want one of the debuggers' I/O to go over an emulated serial port and the other application to use the emulated screen and keyboard. There is no reason to run the dosemu process over a serial port, I just open it (with -t or -dumb) directly from a terminal.

**Reply by stsp** [reply]
So if you open dosemu from terminal, why
is to use -dumb?
Or if the answer is "because we can",
then you can just implement a helper
for dosemu to print directly on terminal,
no matter was it -t or -dumb.

**Reply by ecm-pushbx** [reply]
-dumb is easier to scroll back on the terminal, like when using screen.

**Reply by stsp** [reply]
But not when you will use ANSI on
your own, for sure.
Anyway, I don't see the problem to
implement the raw terminal write
helper function.

**Reply by ecm-pushbx** [reply]
> But not when you will use ANSI on
> your own, for sure.

I don't understand. With -dumb I can use ANSI sequences and they will be handled by the host terminal. This does not diminish the ability to scroll back on the terminal. (NB I want to use only colour/reverse video formatting, no position changes.)

**Reply by stsp** [reply]
Yes, with some limited use, and most
importantly by outputting normal LFs
instead of cursor positioning, you can
keep having scroll-back.
I don't find it exceptionally impossible
to do something similar for -t. For example
if you preserve \r and \n's in -t mode
when int10 uses terminal output funcs
(rather than when we just scan the
VGA text buffer for an updates), then
maybe you can get the scroll-back even
there. But I am not sure of that.
You can try some research on that topic. :)

Or... just implement the raw terminal
output fn and be done.

**Reply by stsp** [reply]
OK, there appeared to be the GET_TERM_TYPE
helper already, which was in a need of an extension
anyway. So I added the dumb_video type there.
But I don't believe you should use it that way.

**Reply by ecm-pushbx** [reply]
> with some limited use, and most
> importantly by outputting normal LFs
> instead of cursor positioning, you can
> keep having scroll-back.

Yes, I only use LFs for linefeed and (repeated) CRs to overwrite existing content on a single line.

**Comment by stsp** [comment]
What I find more productive to discuss is
a way to print to a terminal bypassing int10.
Then you will be able to do your trick w/o
querying the mode.

**Comment by ecm-pushbx** [comment]
I implemented changed register highlighting in lDebug now. When adding the int 10h attribute handling for non-dumb-dosemu environments my tests revealed that interrupt 10h service 03h always returns `dl` = 0 for dosemu2 -dumb. Refer to https://hg.ulukai.org/ecm/ldebug/file/7afd87d4f32c/source/lineio.asm#l803

This would appear to be a way to detect that dosemu -dumb is in use. (I always write at least one byte to a given line before this check so `dl` is always nonzero when not in dumb mode.) It works with older dosemu2 too. What do you think of this method?

**Reply by stsp** [reply]
I think that primitive cursor tracking
is possible in the future even for dumb
mode. For example by checking for
\r and then resetting to zero, otherwise
increment.
